libp2p/protocols/relay.nim

```diff
@@ -0,0 +1,471 @@
+    finally:
+      trace "exiting relay handler", conn
+      await conn.close()
+
+  r.queue = newAsyncQueue[Connection](0)
+  r.handler = handleStream
+  r.codecs = @[RelayCodec]
+
+  r.maxCircuit = MaxCircuit
+  r.maxCircuitPerPeer = MaxCircuitPerPeer
+  r.msgSize = MsgSize
+
+proc dialPeer(
+    r: Relay,
+    conn: Connection,
+    dstPeerId: PeerId,
+    dstAddrs: seq[MultiAddress]): Future[Connection] {.async.} =
+  var
+    msg = RelayMessage(
+      msgType: some(RelayType.Hop),
+      srcPeer: some(RelayPeer(peerId: r.switch.peerInfo.peerId, addrs: r.switch.peerInfo.addrs)),
+      dstPeer: some(RelayPeer(peerId: dstPeerId, addrs: dstAddrs)),
+      status: none(RelayStatus))
+    pb = encodeMsg(msg)
+
+  trace "Dial peer", msgSend=msg
+
+  let msgRcvFromRelayOpt = try:
+    await conn.writeLp(pb.buffer)
+    decodeMsg(await conn.readLp(r.msgSize))
+  except CatchableError as exc:
+    trace "error reading stop response", error=getCurrentExceptionMsg()
+    await sendStatus(conn, RelayStatus.HopCantOpenDstStream)
+    raise exc
+
+  if msgRcvFromRelayOpt.isNone:
+    trace "error reading stop response", msg = msgRcvFromRelayOpt
+    await sendStatus(conn, RelayStatus.HopCantOpenDstStream)
+    raise newException(RelayError, "Hop can't open destination stream")
+
+  let msgRcvFromRelay = msgRcvFromRelayOpt.get()
+  if msgRcvFromRelay.msgType.isNone or msgRcvFromRelay.msgType.get() != RelayType.Status:
+    trace "unexcepted relay stop response", msgType = msgRcvFromRelay.msgType
+    await sendStatus(conn, RelayStatus.HopCantOpenDstStream)
+    raise newException(RelayError, "Hop can't open destination stream")
+
+  if msgRcvFromRelay.status.isNone or msgRcvFromRelay.status.get() != RelayStatus.Success:
+    trace "relay stop failure", status=msgRcvFromRelay.status
+    await sendStatus(conn, RelayStatus.HopCantOpenDstStream)
+    raise newException(RelayError, "Hop can't open destination stream")
+  result = conn
+
+#
+# Relay Transport
+#
+
+const
+  CircuitRelay = mapEq("p2p-circuit")
+  P2PPat = mapEq("p2p")
+
+type
+  RelayTransport* = ref object of Transport
+    relay: Relay
+
+method start*(self: RelayTransport, ma: MultiAddress) {.async.} =
+  if self.running:
+    trace "Relay transport already running"
+    return
+
+  await procCall Transport(self).start(ma)
+  trace "Starting Relay transport"
+
+method stop*(self: RelayTransport) {.async, gcsafe.} =
+  self.running = false
+  while not self.relay.queue.empty():
+    await self.relay.queue.popFirstNoWait().close()
+
+method accept*(self: RelayTransport): Future[Connection] {.async, gcsafe.} =
+  result = await self.relay.queue.popFirst()
+
+# TODO: dial / handles active relay:
+# /ipv4/127.0.0.1/tcp/5002/p2p/QmRelay/p2p-circuit/ipv4/127.0.0.1/tcp/5003/p2p/QmTarget
+# dstPeerId = QmTarget
+# dstAddrs = /ipv4/127.0.0.1/tcp/5003
+#
+# TODO: Must find a way to solve the handles problem, currently the dst switch is created
```

Comment:
Not exactly sure what the problem is, it's should be totally valid to start with the `/p2p-circuit` address, that might be another way of announcing circuit capabilities in the future.

Reply:
Actually, I just re-read this, and I think the issue is that you shouldn't supply the addresses manually at start, since you don't know which relay will relay to you

So I think we shouldn't have any p2p-circuit address in our multiaddress list (until we handle the reservations from circuit relay v2), and should work as-is?

Reply:
Ah, I see - yes, we might allow an empty/no addresses at start, that shouldn't be a big issue.